test/integration/framework/etcd.go

```diff
@@ -128,7 +129,15 @@ func RunCustomEtcd(dataDir string, customFlags []string) (url string, stopFn fun
 	cmd.Stdout = os.Stdout
 	cmd.Stderr = os.Stderr
 	stop := func() {
-		cancel()
+		// try to exit etcd gracefully
+		cmd.Process.Signal(syscall.SIGTERM)
+		go func() {
+			select {
+			case <-ctx.Done():
+			case <-time.After(5 * time.Second):
+				cancel()
```

Comment:
it's not clear to me that `cmd.Process.Signal(syscall.SIGTERM)` will cancel the context

my read is that we need to call `cancel()` unconditionally

Reply:
calling `cmd.Process.Signal(syscall.SIGTERM)` to shut down etcd more gracefully might be fine, but I'm not sure that is letting us hit the `<-ctx.Done()` case... can you add logging and run locally to see if we're always hitting this 5 second timeout case?

Reply:
check the description https://github.com/kubernetes/kubernetes/pull/108048#issue-1130217501

```
Before this patch:

go test -timeout 100s -run ^TestLivezAndReadyz$ k8s.io/kubernetes/test/integration/controlplane -v -count 1

I0210 15:46:15.195847 1877999 etcd.go:133] etcd exit status: signal: killed

After this patch:

I0210 15:48:35.992339 1880623 etcd.go:142] etcd exit status: signal: terminated
```

Reply:
https://github.com/kubernetes/kubernetes/pull/108048/files#r805681021

Reply:
but I see your question, let me double check what happens with the context

Reply:
you are right, the cmd exits but the context is not cancelled
/hold

Reply:
/hold cancel

Cancel() can be called unconditionally https://pkg.go.dev/context#CancelFunc